### PR TITLE
planner: reject invalid conversion from like to =

### LINKED ISSUE #11171: SELECT 0 LIKE 'a string' should not be true.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

```mysql
SELECT 0 LIKE 'a string'
```

2. What did you expect to see?

False with no warning.
```mysql
mysql> SELECT 0 LIKE 'a string';
+-------------------+
| 0 LIKE 'a string' |
+-------------------+
|                 0 |
+-------------------+
1 row in set (0.00 sec)
```

3. What did you see instead?

```mysql
mysql> SELECT 0 LIKE 'a string';
+-------------------+
| 0 LIKE 'a string' |
+-------------------+
|                 1 |
+-------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+----------------+
| Level   | Code | Message        |
+---------+------+----------------+
| Warning | 1265 | Data Truncated |
+---------+------+----------------+
1 row in set (0.01 sec)
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
fdbc149b72db7034d024acf1209e455c0ad4ca9d

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve? 
Fix #11171 

### What is changed and how it works?
Reject  invalid conversion from like to =
`0 like "a"` should not convert to `0 = "a"`,
if so, it will further become `0 = 0` and return `true`
After this pr, `t.col like "a"` can't use indexscan when col isn't a string.
If the col isn't a string, the sql maybe wrong and is ok not to use indexsacn. 

### Check List 

Tests 

 - Integration test

Code changes

Side effects

Related changes

### CODE REVIEW

File: planner/core/expression_rewriter.go [hunk 1]
```diff
@@ -1237,7 +1237,7 @@ func (er *expressionRewriter) patternLikeToExpression(v *ast.PatternLikeExpr) {
 		}
 		if !isNull {
 			patValue, patTypes := stringutil.CompilePattern(patString, v.Escape)
-			if stringutil.IsExactMatch(patTypes) {
+			if stringutil.IsExactMatch(patTypes) && er.ctxStack[l-2].GetType().EvalType().IsStringKind() {
```
Comment (XuHuaiyu) [inline planner/core/expression_rewriter.go]: Is it seems that we should not call `IsStringKind()` but check `.EvalType() == ETString` here?

Comment (wjhuang2016) [reply planner/core/expression_rewriter.go]: > Is it seems that we should not call `IsStringKind()` but check `.EvalType() == ETString` here?

It seems that IsStringKind() is enough, If the val's type is String, they will be compared as string.
https://github.com/pingcap/tidb/blob/5aef053c163524aa167e6bb4ddabd0671a75e712/expression/builtin_compare.go#L958-L978

Comment (XuHuaiyu) [reply planner/core/expression_rewriter.go]: IsStringKind contains ETDatetime, etc.

``` sql
func (et EvalType) IsStringKind() bool {
	return et == ETString || et == ETDatetime ||
		et == ETTimestamp || et == ETDuration || et == ETJson
}
```
> If the val's type is String, they will be compared as string

Noop, `getBaseCmpType` gets a *basic* comparison type, the final comparison type may change later.

Comment (eurekaka) [reply planner/core/expression_rewriter.go]: Though not related to this patch, looks like I found another compatibility issue:
In MySQL:
```
mysql> create table tbl(id int primary key, a date);
Query OK, 0 rows affected (0.03 sec)

mysql> set sql_mode = '';
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql> insert tbl values(0, '0000-00-00');
Query OK, 1 row affected (0.02 sec)

mysql> select a like 'a' from tbl;
+------------+
| a like 'a' |
+------------+
|          0 |
+------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+---------------------------------------------------+
| Level   | Code | Message                                           |
+---------+------+---------------------------------------------------+
| Warning | 1292 | Incorrect date value: 'a' for column 'a' at row 1 |
+---------+------+---------------------------------------------------+
```
while in TiDB (even with this patch):
```
mysql> select a like 'a' from tbl;
+------------+
| a like 'a' |
+------------+
|       NULL |
+------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+-------------------------------+
| Level   | Code | Message                       |
+---------+------+-------------------------------+
| Warning | 1292 | Incorrect datetime value: 'a' |
+---------+------+-------------------------------+
```
@wjhuang2016 are you interested to fix this as well?

File: planner/core/physical_plan_test.go [hunk 2]
```diff
@@ -1170,11 +1170,10 @@ func (s *testPlanSuite) TestRefine(c *C) {
 			sql:  `select a from t where c_str like 123`,
 			best: "IndexReader(Index(t.c_d_e_str)[[\"123\",\"123\"]])->Projection",
 		},
-		// c is type int which will be added cast to specified type when building function signature,
-		// and rewrite predicate like to predicate '='  when exact match , index still can be used.
+		// c is type int which will be added cast to specified type when building function signature, no index can be used.
```
Comment (eurekaka) [inline planner/core/physical_plan_test.go]: We can simply remove this test now.

File: planner/core/expression_rewriter.go (hunk 1 above)
Comment (wjhuang2016) [reply planner/core/expression_rewriter.go]: > IsStringKind contains ETDatetime, etc.
> 
> ```sql
> func (et EvalType) IsStringKind() bool {
> 	return et == ETString || et == ETDatetime ||
> 		et == ETTimestamp || et == ETDuration || et == ETJson
> }
> ```
> 
> > If the val's type is String, they will be compared as string
> 
> Noop, `getBaseCmpType` gets a _basic_ comparison type, the final comparison type may change later.

I got it. If the lhs is a datatime, the rhs may be converted to a datatime.

File: planner/core/physical_plan_test.go (hunk 2 above)
Comment (wjhuang2016) [reply planner/core/physical_plan_test.go]: @eurekaka Yeah, I will fix it.

**Comment by wjhuang2016** [comment]
@eurekaka, after I changed the `IsStringKind()` to `== ETSTRING` as @XuHuaiyu suggested, 
`select a like 'a' from tbl;` return `0` and without warnings;
Besides,  `insert tbl values(0, '0000-00-00');` will exec suceessfully without `set sql_mode = '';` in TiDB

**Comment by eurekaka** [comment]
@wjhuang2016 OK, can you file GitHub issues for those 2 compatibility problems found? Then for this PR, we can focus on the correction of the query result.

**Pull request merged**